storage/src/mmr/mem.rs

```diff
@@ -0,0 +1,854 @@
+//! A bare-bones MMR structure without pruning and where all nodes are hashes & maintained in
+//! memory.
+//!
+//! TERMINOLOGY:
+//!
+//! An MMR is a list of perfect binary trees of strictly decreasing height. The roots of these trees
+//! are called the "peaks" of the MMR.
+//!
+//! Each "element" stored in the MMR is represented by some leaf node in one of these perfect trees.
+//!
+//! The "position" of a node in the MMR is defined as the 0-based index of the node in the
+//! underlying nodes vector. As the MMR is an append-only data structure, node positions never
+//! change and can be used as stable identifiers.
+//!
+//! The "height" of a node is 0 for a leaf, 1 for the parent of 2 leaves, and so on.
+//!
+//! The "root hash" of an MMR is the result of hashing together the size of the MMR and the hashes
+//! of every peak in decreasing order of height.
```

Comment:
nit: think it would be a good idea to include a basic diagram and/or an example of storing some items and computing the root (see cryptography crate for a number of examples)

Reply:
added